src/backend/ingestion/framework-detection.lambda-shared.ts

```diff
@@ -28,65 +40,67 @@ export interface ConstructFramework {
-    if (packageName.startsWith('@aws-cdk/') || packageName === 'aws-cdk-lib' || packageName === 'monocdk') {
-      if (name && name !== ConstructFrameworkName.AWS_CDK) {
-        // Identified multiple candidates, so returning ambiguous...
-        nameAmbiguous = true;
-        return;
-      }
-      name = ConstructFrameworkName.AWS_CDK;
-    } else if (packageName === 'cdktf') {
-      if (name && name !== ConstructFrameworkName.CDKTF) {
-        // Identified multiple candidates, so returning ambiguous...
-        nameAmbiguous = true;
-        return;
-      }
-      name = ConstructFrameworkName.CDKTF;
-    } else if (packageName === 'cdk8s' || /^cdk8s-plus(?:-(?:17|20|21|22))?$/.test(packageName)) {
-      if (name && name !== ConstructFrameworkName.CDK8S) {
-        // Identified multiple candidates, so returning ambiguous...
-        nameAmbiguous = true;
-        return;
-      }
-      name = ConstructFrameworkName.CDK8S;
-    } else {
-      return;
-    }
-    if (versionRange) {
-      const major = minVersion(versionRange)?.major;
-      if (majorVersion != null && majorVersion !== major) {
-        // Identified multiple candidates, so this is ambiguous...
-        majorVersionAmbiguous = true;
+    for (const frameworkName of [ConstructFrameworkName.AWS_CDK, ConstructFrameworkName.CDK8S, ConstructFrameworkName.CDKTF]) {
+      const matchesFramework = FRAMEWORK_MATCHERS[frameworkName];
+      if (matchesFramework(packageName)) {
+        const frameworkVersion: number | 'not-sure' | 'ambiguous' | undefined = detectedFrameworks[frameworkName];
+        const packageMajor: number | 'not-sure' = (versionRange ? minVersion(versionRange)?.major : undefined) ?? 'not-sure';
+
+        if (frameworkVersion === undefined) {
+          // It's the first time seeing this major version, so we record
+          // whatever new information we found ("not-sure" or number).
+          detectedFrameworks[frameworkName] = packageMajor;
+          return;
+        } else if (frameworkVersion === 'ambiguous') {
+          // We have already seen multiple major versions, so just give up
+          // trying to identify this framework.
+          return;
+        } else if (frameworkVersion === 'not-sure') {
+          // We haven't seen a major version for this framework yet, so record
+          // whatever new information we found ("not-sure" or number).
+          detectedFrameworks[frameworkName] = packageMajor;
+          return;
+        } else if (typeof frameworkVersion === 'number') {
+          // We've seen evidence of a particular major version for this
+          // framework, so only update if this package conflicts with what
+          // we're expecting.
+          if (packageMajor !== 'not-sure' && frameworkVersion !== packageMajor) {
+            detectedFrameworks[frameworkName] = 'ambiguous';
+          }
+          return;
+        }
```

Comment:
I'd argue there's a missing `else` clause here... Supposedly, when you reach that final `else if`, `frameworkVersion` is guaranteed to be a number...

So Maybe switch to:

```suggestion
        } else {
          // At this point, frameworkVersion can only be a MV number
          assert(typeof frameworkVersion === 'number');
          
          // We've seen evidence of a particular major version for this
          // framework, so only update if this package conflicts with what
          // we're expecting.
          if (packageMajor !== 'not-sure' && frameworkVersion !== packageMajor) {
            detectedFrameworks[frameworkName] = 'ambiguous';
          }
          return;
        }
```